File: src/main/java/org/blacksun/graph/channel/AbstractChannel.java
package org.blacksun.graph.channel;

import org.blacksun.graph.node.GraphNode;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

abstract class AbstractChannel implements Channel {
    protected final GraphNode fromNode;
    protected final GraphNode toNode;
    protected boolean used;
    protected int weight;
    protected double errors;

    public AbstractChannel(@NotNull GraphNode fromNode, @NotNull GraphNode toNode, int weight, double errors) {
        this.fromNode = fromNode;
        this.toNode = toNode;
        this.weight = weight;
        this.errors = errors;
        used = false;
    }

    protected abstract String getDirection();

    @Override
    public GraphNode getFromNode() {
        return fromNode;
    }

    @Override
    public GraphNode getToNode() {
        return toNode;
    }

    @Override
    public void setWeight(int weight) {
        this.weight = weight;
    }

    @Override
    public int getWeight() {
        return weight;
    }

    @Override
    public void setErrors(double errors) {
        this.errors = errors;
    }

    @Override
    public double getErrors() {
        return errors;
    }

    @Override
    public void setUsed(boolean used) {
        this.used = used;
    }

    @Override
    public boolean isUsed() {
        return used;
    }

    @Override
    public void connect() {
        fromNode.addConnection(this);
    }

    @Override
    public void remove() {
        fromNode.removeConnection(this);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        AbstractChannel that = (AbstractChannel) o;
        return used == that.used &&
                weight == that.weight &&
                Double.compare(errors, that.errors) == 0 &&
                Objects.equals(fromNode, that.fromNode) &&
                Objects.equals(toNode, that.toNode);
    }

    @Override
    public int hashCode() {
        return Objects.hash(fromNode, toNode, used, weight, errors);
    }

    @Override
    public String toString() {
        return String.format("%s %s %s (weight=%d, errors=%.1f%%)",
                fromNode, getDirection(), toNode, weight, errors * 100);
    }
}

File: src/main/java/mvc/model/channel/AbstractChannel.kt
package mvc.model.channel

import mvc.model.node.Node

import java.util.Objects

abstract class AbstractChannel(override val fromNode: Node, override val toNode: Node, override var weight: Int, override var errors: Double) : Channel {
    override var isUsed = false
    override var isSelected = false

    protected abstract val direction: String

    override fun connect() {
        fromNode.addConnection(this)
    }

    override fun remove() {
        fromNode.removeConnection(this)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as AbstractChannel?
        return isUsed == that!!.isUsed &&
                weight == that.weight &&
                java.lang.Double.compare(errors, that.errors) == 0 &&
                fromNode == that.fromNode &&
                toNode == that.toNode
    }

    override fun hashCode(): Int {
        return Objects.hash(fromNode, toNode, isUsed, weight, errors)
    }

    override fun toString(): String {
        var busy = ""
        if (isUsed) {
            busy = ", busy"
        }
        return String.format("%s %s %s (weight=%d, errors=%.1f%%%s)",
                fromNode, direction, toNode, weight, errors * 100, busy)
    }
}
